manifests/runner/docker.pp

```diff
@@ -0,0 +1,186 @@
+    [$cap_drop, $gitlab::cirunner::docker::default_cap_drop]
+  )
+  $security_opt_cmd = ::gitlab::cirunner::cmd_str(
+    'docker-security-opt',
+    [$security_opt, $gitlab::cirunner::docker::default_security_opt]
+  )
+  $devices_cmd = ::gitlab::cirunner::cmd_str(
+    'docker-devices',
+    [$devices, $gitlab::cirunner::docker::default_devices]
+  )
+  $disable_cache_cmd = ::gitlab::cirunner::cmd_str(
+    'docker-disable-cache',
+    [$disable_cache, $gitlab::cirunner::docker::default_disable_cache]
+  )
+  $volume_driver_cmd = ::gitlab::cirunner::cmd_str(
+    'docker-volume-driver',
+    [$volume_driver, $gitlab::cirunner::docker::default_volume_driver]
+  )
+  $extra_hosts_cmd = ::gitlab::cirunner::cmd_str(
+    'docker-extra-hosts',
+    [$extra_hosts, $gitlab::cirunner::docker::default_extra_hosts]
+  )
+  $volumes_from_cmd = ::gitlab::cirunner::cmd_str(
+    'docker-volumes-from',
+    [$volumes_from, $gitlab::cirunner::docker::default_volumes_from]
+  )
+  $network_mode_cmd = ::gitlab::cirunner::cmd_str(
+    'docker-network-mode',
+    [$network_mode, $gitlab::cirunner::docker::default_network_mode]
+  )
+  $links_cmd = ::gitlab::cirunner::cmd_str(
+    'docker-links',
+    [$links, $gitlab::cirunner::docker::default_links]
+  )
+  $services_cmd = ::gitlab::cirunner::cmd_str(
+    'docker-services',
+    [$services, $gitlab::cirunner::docker::default_services]
+  )
+  $wait_for_services_timeout_cmd = ::gitlab::cirunner::cmd_str(
+    'docker-wait-for-services-timeout',
+    [$wait_for_services_timeout, $gitlab::cirunner::docker::default_wait_for_services_timeout]
+  )
+  $allowed_images_cmd = ::gitlab::cirunner::cmd_str(
+    'docker-allowed-images',
+    [$allowed_images, $gitlab::cirunner::docker::default_allowed_images]
+  )
+  $allowed_services_cmd = ::gitlab::cirunner::cmd_str(
+    'docker-allowed-services',
+    [$allowed_services, $gitlab::cirunner::docker::default_allowed_services]
+  )
+  $pull_policy_cmd = ::gitlab::cirunner::cmd_str(
+    'docker-pull-policy',
+    [$pull_policy, $gitlab::cirunner::docker::default_pull_policy]
+  )
+  $shm_size_cmd = ::gitlab::cirunner::cmd_str(
+    'docker-shm-size',
+    [$shm_size, $gitlab::cirunner::docker::default_shm_size]
+  )
+  $parameters_string = [
+    $executor_cmd, $token_cmd, $url_cmd, $name_cmd, $run_untagged_cmd, $locked_cmd,
+    $tags_cmd, $host_cmd, $tlsverify_cmd, $hostname_cmd, $image_cmd,
+    $cpuset_cpus_cmd, $cpus_cmd, $dns_cmd, $dns_search_cmd, $privileged_cmd,
+    $userns_cmd, $cap_add_cmd, $cap_drop_cmd, $security_opt_cmd, $devices_cmd,
+    $disable_cache_cmd, $volume_driver_cmd, $extra_hosts_cmd, $volumes_from_cmd,
+    $network_mode_cmd, $links_cmd, $services_cmd, $wait_for_services_timeout_cmd,
+    $allowed_images_cmd, $allowed_services_cmd, $pull_policy_cmd, $shm_size_cmd,
+  ].join(' ')
+  # Execute gitlab ci multirunner register
+  exec {"Register_runner_${title}":
+    command => "/usr/bin/gitlab-ci-multi-runner register ${parameters_string}",
```

Comment:
should we use https://github.com/puppetlabs/puppetlabs-stdlib#shell_escape or https://github.com/puppetlabs/puppetlabs-stdlib#shell_join here?

Reply:
I could probably replace the join used above with the shell_join function.  however this variable has no user submitted data so it should be safe

Reply:
i can change to shell_join i think

Reply:
@bastelfreak i looked at using both shell_join and shell_escape however neither of theses produced the desired command.

### shell_join produced
`/usr/bin/gitlab-ci-multi-runner register --executor\\ docker --registration-token\\ TOKEN --url\\ https://gitlab.com -n\\ --name\\ docker_docker_runner --run-untagged '' '' --docker-host\\ foobar '' '' --docker-image\\ ubuntu_trusty '' '' '' '' '' '' '' '' '' '' '' '' '' '' '' '' '' '' '' '' '' ''`

### join(' ').shell_escape produced
`/usr/bin/gitlab-ci-multi-runner register --executor\\ docker\\ --registration-token\\ TOKEN\\ --url\\ https://gitlab.com\\ -n\\ --name\\ docker_docker_runner\\ --run-untagged\\ \\ \\ --docker-host\\ foobar\\ \\ \\ --docker-image\\ ubuntu_trusty\\ \\ \\ \\ \\ \\ \\ \\ \\ \\ \\ \\ \\ \\ \\ \\ \\ \\ \\ \\ \\ \\`